enclave-runtime/src/sidechain_impl.rs

```diff
@@ -0,0 +1,268 @@
+	LightClient: Validator<PB> + Send + Sync + 'static,
+	Author: AuthorApi<H256, PB::Hash>
+		+ SendState<Hash = PB::Hash>
+		+ OnBlockCreated<Hash = PB::Hash>
+		+ Send
+		+ Sync,
+{
+	type Proposer = SlotProposer<PB, SB, P, OcallApi, LightClient, Author>;
+	type Error = ConsensusError;
+
+	fn init(
+		&mut self,
+		parent_header: &PB::Header,
+		shard: ShardIdentifierFor<SB>,
+	) -> Result<Self::Proposer, Self::Error> {
+		Ok(SlotProposer {
+			ocall_api: self.ocall_api.clone(),
+			light_client: self.light_client.clone(),
+			author: self.author.clone(),
+			proposer_key: self.pair.clone(),
+			parentchain_header: parent_header.clone(),
+			shard,
+			_phantom: PhantomData,
+		})
+	}
+}
+
+impl<PB, SB, Pair, OcallApi, LightClient, Author> Proposer<PB, SB>
+	for SlotProposer<PB, SB, Pair, OcallApi, LightClient, Author>
+where
+	PB: Block<Hash = H256>,
+	NumberFor<PB>: BlockNumberOps,
+	SB: SignedBlock<Public = sp_core::ed25519::Public, Signature = MultiSignature>,
+	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
+	OcallApi: EnclaveOnChainOCallApi,
+	LightClient: Validator<PB> + Send + Sync + 'static,
+	Author:
+		AuthorApi<H256, PB::Hash> + SendState<Hash = PB::Hash> + OnBlockCreated<Hash = PB::Hash>,
+{
+	fn propose(&self, max_duration: Duration) -> Result<Proposal<SB>, ConsensusError> {
+		let (calls, blocks) = exec_tops::<PB, SB, _, _>(
+			self.ocall_api.as_ref(),
+			self.author.as_ref(),
+			&self.parentchain_header,
+			self.shard,
+			max_duration,
+		)
+		.map_err(|e| ConsensusError::Other(e.to_string().into()))?;
+
+		Ok(Proposal {
+			block: blocks.ok_or(ConsensusError::CannotPropose)?,
+			parentchain_effects: calls,
+		})
+	}
+}
+
+/// Executes aura for the given `slot`
+pub fn exec_aura_on_slot<Authority, PB, SB, OcallApi, LightValidator, Author>(
+	slot: SlotInfo<PB>,
+	authority: Authority,
+	rpc_author: Author,
+	validator: &mut LightValidator,
+	ocall_api: OcallApi,
+	nonce: &mut u32,
+	shards: Vec<ShardIdentifierFor<SB>>,
+) -> EnclaveResult<()>
+where
+	// setting the public type is necessary due to some non-generic downstream code.
+	Authority: Pair<Public = sp_core::ed25519::Public>,
+	Authority::Public: Encode,
+	PB: Block<Hash = H256>,
+	SB: SignedBlock<Public = Authority::Public, Signature = MultiSignature> + 'static,
+	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = Authority::Public>,
+	OcallApi: EnclaveOnChainOCallApi + 'static,
+	LightValidator: Validator<PB> + LightClientState<PB> + Clone + Send + Sync + 'static,
+	NumberFor<PB>: BlockNumberOps,
+	Author: AuthorApi<H256, PB::Hash>
+		+ SendState<Hash = PB::Hash>
+		+ OnBlockCreated<Hash = PB::Hash>
+		+ Send
+		+ Sync,
+{
+	log::info!("[Aura] Executing aura for slot: {:?}", slot);
+
+	let env = ProposerFactory::new(
+		Arc::new(ocall_api.clone()),
+		Arc::new(validator.clone()),
+		Arc::new(rpc_author),
+		authority.clone(),
+	);
+
+	let mut aura = Aura::<_, _, SB, _, _>::new(authority, ocall_api.clone(), env)
```

Comment:
So aura takes an `ocall_api`, but also an `env` which contains the `ocall_api` ?

Reply:
Good catch 👍 . I changed the design at some point and did not do a proper cleanup!

Reply:
Ah, I remember now. I was thinking about changing the `env`'s `init` function to take `ocall_api` as parameter. I decided against it as the business logic specific `proposer` does not always need to have an `OcallApi`.

Reply:
I suspected something like that, happens to me as well, all the time (especially in large change sets like this one is 😄 )